kernel-rs/src/fs/lfs/inode.rs

```diff
@@ -332,51 +394,50 @@ impl Itable<Lfs> {
         tx: &Tx<'_, Lfs>,
         ctx: &KernelCtx<'_, '_>,
     ) -> RcInode<Lfs> {
-        loop {
-            let segment = &tx.fs.find_segment();
-            let mut bp = hal().disk().read(dev, segment.offset, ctx);
-
-            for inum in 1..tx.fs.superblock().ninodes {
-                const_assert!(IPB <= mem::size_of::<BufData>() / mem::size_of::<Dinode>());
-                const_assert!(mem::align_of::<BufData>() % mem::align_of::<Dinode>() == 0);
-                // SAFETY: dip is inside bp.data.
-                let dip = unsafe {
-                    (bp.deref_inner_mut().data.as_mut_ptr() as *mut Dinode).add(inum as usize % IPB)
-                };
-                // SAFETY: i16 does not have internal structure.
-                let t = unsafe { *(dip as *const i16) };
-                // If t >= #(variants of DInodeType), UB will happen when we read dip.typ.
-                assert!(t < core::mem::variant_count::<DInodeType>() as i16);
-                // SAFETY: dip is aligned properly and t < #(variants of DInodeType).
-                let dip = unsafe { &mut *dip };
-
-                // a free inode
-                if dip.typ == DInodeType::None {
-                    // SAFETY: DInode does not have any invariant.
-                    unsafe { memset(dip, 0u32) };
-                    match typ {
-                        InodeType::None => dip.typ = DInodeType::None,
-                        InodeType::Dir => dip.typ = DInodeType::Dir,
-                        InodeType::File => dip.typ = DInodeType::File,
-                        InodeType::Device { major, minor } => {
-                            dip.typ = DInodeType::Device;
-                            dip.major = major;
-                            dip.minor = minor;
-                        }
-                    }
-
-                    // mark it allocated on the disk
-                    tx.write(bp, ctx);
-                    return self.get_inode(dev, inum);
-                } else {
-                    bp.free(ctx);
-                }
+        let mut imap = tx.fs.imap();
+        let mut segment = tx.fs.segment();
+
+        // 1. Write the inode.
+        let inum = imap.get_empty_inum(ctx).unwrap();
+        let (mut bp, disk_block_no) = segment.add_new_inode_block(inum, ctx).unwrap();
+
+        const_assert!(IPB <= mem::size_of::<BufData>() / mem::size_of::<Dinode>());
+        const_assert!(mem::align_of::<BufData>() % mem::align_of::<Dinode>() == 0);
+        // SAFETY: dip is inside bp.data.
+        let dip = unsafe {
+            (bp.deref_inner_mut().data.as_mut_ptr() as *mut Dinode).add(inum as usize % IPB)
```

Comment:
* I'm planning to place only one inode on each disk block. (i.e. The `inum as usize % IPB` part is just a mistake. It should be changed to `inum as usize`. Again, sorry for the mistake and thank you for mentioning it.)
  * I insist on doing it this way since other papers (ex: https://pages.cs.wisc.edu/~remzi/OSTEP/file-lfs.pdf) and implementations did it this way. Doing it the other way may make things too complicated. However, I think we could change it later after we finally implement the lfs.